File: solutions/javascript/parallel-letter-frequency/parallel-letter-frequency.js
export const parallelLetterFrequency = (texts) => {
  const letterFrequency = (frequencies, text) => {
    return new Promise((resolve) => {
      for (const [letter] of text.toLowerCase().match(/\p{Letter}/gu) || []) {
        frequencies[letter] = (frequencies[letter] || 0) + 1;
      }
      resolve(frequencies);
    });
  };

  // Promises provide concurency but not true parallelism
  const result = {};
  Promise.all(texts.map((text) => letterFrequency(result, text)));
  return new Promise((resolve) => resolve(result));
};

File: solutions/csharp/parallel-letter-frequency/ParallelLetterFrequency.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class ParallelLetterFrequency
{
    public static Dictionary<char, int> Calculate(IEnumerable<string> texts) =>
        texts.AsParallel().Aggregate(new Dictionary<char, int>(), letterFrequency);

    private static Dictionary<char, int> letterFrequency(Dictionary<char, int> frequencies, string text) =>
        text.ToLower().Where(char.IsLetter).Aggregate(frequencies, addLetterFrequency);

    private static Dictionary<char, int> addLetterFrequency(Dictionary<char, int> frequency, char letter)
    {
        frequency.TryAdd(letter, 0);
        frequency[letter] += 1;
        return frequency;
    }
}